### ISSUE TITLE
EPIC 9 — Persistent Visual Scene Memory and Vision-Based Avoidance

### ISSUE THREAD

**Issue by profhamilton3** [profhamilton3/reachy-tabletop-ai#13]
## 1. Purpose

Reachy can load a scene and physics model in `reachy-1-2-sim` (ground-truth
geometry) and can see the real world through its stereo head cameras, but
nothing today lets those two facts build on each other. There is no persisted
model of "what Reachy has seen" — no scene/object memory that survives a
restart, and no path from "the camera sees the rail edge and table edge" to
"estimate and avoid before moving."

Epic 9 builds that link: a **persistent visual scene/object memory** that
survives reboots and grows with use, trained and validated first in
simulation (where ground-truth geometry gives free, perfect supervision)
before any of it is trusted against the real cameras.

## 2. Expected outcome

At Epic 9 completion:

1. a lightweight detector can locate structural boundaries (rail edge, table
   edge) and known objects in a camera frame;
2. that detector was trained entirely from simulator renders labeled
   automatically from the compiled scene's own geometry — no manual bounding
   boxes for structural edges;
3. detector output is converted into a `SceneModel`-compatible obstacle
   estimate and checked against the same collision-validation path the
   Cartesian planner already uses (`reachy-1-2-sim`'s
   `src/reachy_ai/scene/awareness.py::SceneModel` /
   `src/reachy_ai/motion/kinematics.py`), so a move can be rejected *before*
   any joint command is sent, from vision alone;
4. confirmed visual/object knowledge persists in a versioned local store that
   survives a process restart and accumulates across sessions, instead of
   resetting to nothing every run;
5. persisted knowledge is invalidated, not silently trusted, when the
   physical rig or calibration profile it was learned under has changed;
6. every step of this is validated in `reachy-1-2-sim` against ground truth
   before being pointed at the real robot's cameras.

## 3. Relationship to `reachy-1-2-sim`

The training and validation loop lives in `reachy-1-2-sim`, which already
owns the pieces this depends on:

- ground-truth scene geometry (table, rails, console) via
  `native_mujoco/scene_compiler.py`;
- `src/reachy_ai/scene/awareness.py::SceneModel` — table surface, static
  obstacles, panel obstacles, already consumed by the Cartesian planner for
  collision validation against *known* geometry;
- `src/reachy_ai/evaluation/contacts.py` — forbidden-contact / clearance
  checking against ground-truth physics state;
- the `EPIC-8.md` provenance discipline (`SimulatorIdentity`,
  `ExperienceStore`) — a versioned, fail-closed, SQLite-backed pattern for
  "does this persisted knowledge still apply to the world in front of me,"
  which Epic 9's scene-memory store should mirror rather than reinvent.

`reachy-1-2-sim` should own: rendering labeled training data from scene
geometry, training/validating the detector against that ground truth, and
exposing a read-only "does this vision estimate match ground truth" scoring
API.

`reachy-tabletop-ai` (this repo) should own: the real camera pipeline
(`src/reachy_ai/perception/`, currently an empty stub), the persistent scene
memory store, and the avoidance precheck that gates real motion — consuming
the sim-validated detector, not duplicating the training loop.

## 4. Non-goals

Epic 9 does **not** include:

- calling a cloud VLM per frame for obstacle detection — `CLAUDE.md` already
  forbids this; the planner remains event-level only, and this epic's
  detector is a local, lightweight model, not a VLM call;
- emitting raw joint angles or trajectories from the vision system — any
  avoidance decision still routes through `motion/primitives.py` and
  `safety.gate_check()` like every other move;
- full SLAM / dense 3D reconstruction of the lab. Structural-edge and
  known-object detection first; dense mapping is out of scope until this is
  proven;
- neural-network policies in the real-time control loop — avoidance is a
  precheck before a bounded motion primitive runs, not a continuous
  vision-in-the-loop controller;
- replacing `reachy-1-2-sim`'s Epic 8 trajectory-recipe/search work. Epic 9
  is a separate, complementary memory: *what the world looks like*, not
  *what motion worked*.

### 5.1 Ground truth first, real camera second

Nothing trained under this epic is pointed at a real camera until it has
been scored against simulator ground truth (known edge/plane position from
the compiled MJCF, known object pose from the scene YAML) and passes a
documented accuracy threshold. This mirrors Epic 8's "fast search, faithful
validation" split (direct/native path vs. compatibility-replay path) applied
to vision instead of motion.

### 5.2 Fail closed on stale or unfamiliar knowledge

Persisted scene/object knowledge is reusable only when it was learned under
a matching rig/calibration identity (parallel to `SimulatorIdentity`'s
`matches()`). A moved rail, a re-calibrated camera, or a changed lab layout
must invalidate the affected memory rather than silently keep serving it.
Unrecognized or stale state must degrade to "ask/refuse," never to "guess."

### 5.3 Avoidance is a precheck, not a controller

"Estimate an avoidance route before moving" means: run the vision estimate,
build/update a `SceneModel`-compatible obstacle set, run it through the
existing collision-validation path, and only then hand a (possibly
adjusted) plan to `motion/primitives.py`. It is not a per-frame reactive
loop and does not bypass `safety.gate_check()`.

### 5.4 Grows, doesn't drift

New confirmed observations should refine confidence in existing scene/object
knowledge (e.g. repeated confirmation of the table edge position), not
overwrite it on a single noisy frame. Every stored fact keeps enough
provenance (when observed, under what identity, what confidence) to support
this without the store becoming an unaudited black box.

## 6. Proposed data contracts

These should be immutable dataclasses with explicit schema versions, in the
style of `reachy-1-2-sim`'s Section 7 contracts.

### 6.1 `RigIdentity` (parallel to `SimulatorIdentity`)

```text
identity_version
rig_layout_id            # e.g. FWDCenterLabMCC
calibration_profile_id
camera_intrinsics_hash
last_verified_at
```

### 6.2 `VisualObservation`

```text
observation_id
rig_identity
timestamp
source                   # sim_render | real_camera
frame_reference          # camera id / pose at capture time
detected_class           # rail_edge | table_edge | object.<type> | ...
estimated_geometry       # plane/edge/bbox, in robot frame
confidence
ground_truth_error       # populated only for sim-sourced observations
```

### 6.3 `SceneMemoryEntry`

```text
entry_id
rig_identity
semantic_class
geometry_estimate         # current best estimate
confirmation_count
last_confirmed_at
first_observed_at
superseded_by             # set when a later, higher-confidence entry replaces this one
```

## 7. Evaluation model

A detector or scene-memory entry is trustworthy only when:

1. it was scored against simulator ground truth (position/edge error within
   a documented tolerance) before ever running on a real camera;
2. its `RigIdentity` matches the identity of the world currently in front of
   the robot;
3. an avoidance decision built from it can be explained (which entries,
   which confidence, which rejected/accepted move) — no opaque "vision said
   no."

Lexicographic ranking for any avoidance decision, mirroring Epic 8 Section
8.4:

```text
valid rig identity / fresh observation
  > safe (no estimated collision with known or observed geometry)
  > task success
  > accuracy of the avoidance margin
  > smoothness / minimal detour
```

## 8. Suggested phased work

Each phase should close with the same discipline Epic 8 PRs use: stop after
each phase, report what was actually run vs. only unit-tested, and defer
anything not explicitly in scope.

### Phase 0 — Contracts and store

- Implement `RigIdentity`, `VisualObservation`, `SceneMemoryEntry`.
- Add a persistent store (SQLite, mirroring `ExperienceStore`'s
  transactional/versioned pattern) in this repo.
- No detector yet.

### Phase 1 — Sim-rendered, ground-truth-labeled structural detector

- In `reachy-1-2-sim`: render labeled frames for rail edge / table edge
  directly from compiled scene geometry (no manual annotation).
- Train a lightweight edge/plane detector; score it against the same
  ground truth it was rendered from.
- Exit gate: documented accuracy (position/edge error) on held-out
  sim-rendered frames.

### Phase 2 — Vision-derived `SceneModel` precheck

- Convert detector output into a `SceneModel`-compatible obstacle estimate.
- Run it through the existing collision-validation path
  (`scene/awareness.py`, consumed by `motion/kinematics.py`) as a precheck
  before a bounded motion primitive executes.
- Exit gate: a deliberately-placed sim obstacle is detected from rendered
  camera frames and correctly blocks/adjusts a planned move, matching the
  ground-truth collision check.

### Phase 3 — Persistence and growth

- Wire confirmed observations into `SceneMemoryEntry` accumulation
  (confidence growth on repeated confirmation, not overwrite-on-noise).
- Verify state survives a process restart and is correctly invalidated when
  `RigIdentity` no longer matches.

### Phase 4 — Real-camera transfer

- Point the sim-validated detector at the real stereo cameras
  (`src/reachy_ai/perception/`, currently empty).
- Re-run the Phase 1 accuracy scoring against real, measured geometry
  (the FWD Center lab's measured rail/table dimensions) rather than sim
  ground truth.
- Only after this passes: enable the Phase 2 precheck against real camera
  input, gated as always by `safety.gate_check()` and
  `REACHY_ENABLE_MOTION`.

## 9. Open questions to resolve before Phase 0

- Where should the persistent store actually live — this repo, or should it
  be a read-only consumer of something `reachy-1-2-sim` exposes (same
  question Epic 8 leaves open for its own experience-query API)?
- What accuracy tolerance counts as "trustworthy enough to gate real motion"
  for edge/plane position, and who signs off on it (mirrors Epic 8's
  human-approved promotion stage)?
- Does `RigIdentity` need to track camera zoom level separately, given
  `docs/CameraCalibrationFindings` notes that stereo calibration is only
  valid for the zoom level it was captured at?

## Source

Follows from discussion in this repo's issue tracker on universal
scene/object learning that should persist and grow across restarts, and
vision-based avoidance using the rail/table edges Reachy can already see.
Builds directly on `reachy-1-2-sim`'s `EPIC-8.md` provenance/persistence
pattern and on issue #12 (custom object-detection training pipeline).
